### PR TITLE
fix for NoSuchElementException impacting performance

### LINKED ISSUE #35: lots of NoSuchElementException
we're hitting https://github.com/playframework/netty-reactive-streams/blob/master/netty-reactive-streams-http/src/main/java/com/typesafe/netty/http/HttpStreamsServerHandler.java#L221-L229 almost all of the time in our app, resulting in several exceptions per user request. Not great for performance, as you can imagine.

Can you think of a guard that could be performed before the `remove` to ensure the method (and exception) is not called unnecessarilly?

### PR DESCRIPTION
**Pull request by fommil**
close #35 

can you please cherry pick to `master`?

### CODE REVIEW

**Comment by fommil** [comment]
failures look unrelated

File: netty-reactive-streams-http/src/main/java/com/typesafe/netty/http/HttpStreamsServerHandler.java [hunk 1]
```diff
@@ -222,7 +222,9 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
         super.handlerRemoved(ctx);
         for (ChannelHandler dependent: dependentHandlers) {
             try {
-                ctx.pipeline().remove(dependent);
+                // guard to hopefully avoid unnecessary exceptions
+                if (ctx.pipeline().context(dependent) != null)
+                    ctx.pipeline().remove(dependent);
             } catch (NoSuchElementException e) {
```
Comment (viktorklang) [inline netty-reactive-streams-http/src/main/java/com/typesafe/netty/http/HttpStreamsServerHandler.java]: Is the try-catch still needed after the proposed changes?

Comment (fommil) [reply netty-reactive-streams-http/src/main/java/com/typesafe/netty/http/HttpStreamsServerHandler.java]: I'd keep it just incase the context exists momentarily and then is removed in the meantime... but I'm not seeing it being caught anymore.

Comment (viktorklang) [reply netty-reactive-streams-http/src/main/java/com/typesafe/netty/http/HttpStreamsServerHandler.java]: Does the test suite pass w/o it?

Comment (fommil) [reply netty-reactive-streams-http/src/main/java/com/typesafe/netty/http/HttpStreamsServerHandler.java]: I've not see the test suite pass in any situation... there are some flakey tests :man_shrugging: 

I doubt the test suite would be advanced enough to catch the race condition. I've been hundreds of thousands of user requests as part of perftests recently and not seen a single exception... but who's to say it won't happen when I'm not looking?

**Comment by fommil** [comment]
closing for lack of interest

**Pull request closed without merging**